src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -110,7 +117,7 @@ Score(const size_t queryIndex, TreeType& referenceNode)
   if (tree::TreeTraits<TreeType>::FirstPointIsCentroid &&
       lastQueryIndex == queryIndex &&
       traversalInfo.LastReferenceNode() != NULL &&
-      traversalInfo.LastReferenceNode()->Point(0) == referenceNode.Point(0))
+      lastReferenceIndex == referenceNode.Point(0))
```

Comment:
I didn't find that this made a difference---did you observe that `traversalInfo.LastReferenceNode().Point(0)` was sometimes different than `lastReferenceIndex`?

Reply:
Yes, let's say we are using a binary tree:
1. The function that updates LastReferenceNode() is score.
Our tree:
```
    ...
   /
   A
 /   \
B     C
```
Lets say we are going to traverse A.
The traversal behaviour is:
```
Score(B) -> LastNode = B
Score(C) -> LastNode = C

if (Score(B) < Score(C))
  Traverse(B) -> LastReferenceIndex = Whatever there's in B.
...
```
And actually this change fixed `kernelMin < kernelEstimationValue < kernelMax` not being true :)